### ISSUE TITLE
F2F tutorial: pacman image cannot be loaded onto CARTA

### ISSUE THREAD

**Issue by tnakazato** [casangi/xradio#45]
The pacman CASA image created by `write_image(pacman, "pacman_casa.im", out_format="casa")` in image_tutorial.ipynb cannot be loaded onto CARTA nor CASAViewer. If you open it with image analysis tool, it also failed.

```
CASA <8>: ia.open('pacman_casa.im')
2023-10-25 18:08:33     SEVERE  image::open (file src/tools/image/image_cmpt.cc, line 4208)     Exception Reported: Wrong QuantumHolder to convert asQuantumDouble
```

It turned out that imageinfo holds wrong per-plane beams. Each beam parameter should be quantity with single value. But its value is numpy array in the case of pacman image. See actual output below.

```
# get imageinfo from new_casa.im
tb.open('new_casa.im/')
new_casa_imageinfo = tb.getkeyword('imageinfo')
tb.close()

# get imageinfo from pacman_casa.im
tb.open('pacman_casa.im')
pacman_imageinfo = tb.getkeyword('imageinfo')
tb.close()

import pprint
pprint.pprint(new_casa_imageinfo['perplanebeams']['*0'])
# this is valid output
{'major': {'unit': 'rad', 'value': 0.001454441043328608},
 'minor': {'unit': 'rad', 'value': 0.0011635528346628863},
 'positionangle': {'unit': 'rad', 'value': 0.3490658503988659}}

pprint.pprint(pacman_imageinfo['perplanebeams']['*0'])
# this is the output from pacman image
{'major': {'unit': 'rad',
           'value': array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       ...,
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan]])},
 'minor': {'unit': 'rad',
           'value': array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       ...,
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan]])},
 'positionangle': {'unit': 'rad',
                   'value': array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       ...,
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan]])}}
```

So, I think something weird happened when xradio wrote imageinfo for pacman image to disk. If you replace imageinfo with valid one taken from new_casa.im, you can load pacman image onto CARTA. 

Another finding is that pacman is rotated 90 degree left compared with the figure in the tutorial notebook. I don't know if this is a real problem (such as mix-up of coordinates) or just a superficial issue.

![pacman_casa2 im-image-2023-10-26-03-19-13](https://github.com/casangi/xradio/assets/17893104/fa2ab6cd-ab89-4cba-9a6d-7cad19dbefc3)

**Comment by tnakazato** [comment]
FYI, tutorial notebook is in astroviper repo.

https://github.com/casangi/astroviper/blob/main/face_2_face/image_tutorial.ipynb

**Comment by dmehring** [comment]
Thanks very much for the report @tnakazato . At least part of the problem appears to be that xds.where() is unexpectedly (IMO) adding dimensions to the beam array.
```
l_shift = xds.l - len(xds.l)/2
m_shift = xds.m - len(xds.m)/2
# select a circular region of radius 40 pixels
where_sel = xds.where(l_shift**2 + m_shift**2 <= 1600)
where_sel
```
xds.beam has the expected dimensions and shape: (time, polarization, frequency, beam_param) and (1, 4, 50, 3). On the other hand, where_sel.beam has dimensions (time, polarization, frequency, beam_param, l, m) and shape (1, 4, 50, 3, 200, 100). And most, if not all values in the array are nan. (l and m) are dimensions but do not have associated dimension coordinates (yet).

@Jan-Willem do you expect this behavior? The selection is on dimensions (l, m), however, the initial beam does not depend on either of those, but after the where() call, the beam does depend on them and has been expanded so that the (l, m) plane is now the shape of the (l, m) plane in the image.

**Comment by dmehring** [comment]
ok, it looks like I'm thinking of where() wrong. It doesn't do a selection, it creates a mask and behaves the same way xarray.Dataset binary math functions do. https://docs.xarray.dev/en/stable/user-guide/computation.html#comput

I have come up with a solution, but it is not pretty and I'm guessing there is a better and easier way to do it. But anyway, an example looks like

```
# the next line chooses one (unbracketed) value of l and m where where() is true, 
# and so there are non-nans for the beam value. specifying like this implicitly
# drops l and m from nb
nb = where_sel.beam.isel(l=110,m=60)
where_sel.drop_vars('beam')
where_sel['beam'] = nb
where_sel
```

**Comment by dmehring** [comment]
I believe this issue is now fixed on main HEAD. What I do now is, instead of calling where() on the xarray.Dataset, I instead call it on the sky data variable only. This makes since, because that's the only data variable we want to modify; calling where() on the Dataset affects all data variables. In particular, it adds unwanted dimensions to beam.

I also think I understand the rotation. Note that, in the pacman image in the test, l (~RA) is plotted along the y-axes and m (~Dec) is plotted on the x-axis. Your image is plotted in the normal fashion with those axes swapped. Hence the apparent rotation is do to swapping of plotted axes.

In the notebook, I now write the pacman image to a casa image and plot the result (after some transformations) and it agrees with the XDS version. Please have a look if you have time.

Thanks again for the report. It was definitely a learning experience.

**Comment by tnakazato** [comment]
Thanks @dmehring for the update. I definitely want to try updated notebook. Have you pushed your change to remote repo? Is that same notebook as the one we used for hands-on session?

**Comment by dmehring** [comment]
Hi @tnakazato . Yes, changes are on the main branch of xradio. It is the same notebook. Exporting the pacman image in CASA format is done at the end of the notebook. Thanks in advance for having a look!

**Comment by tnakazato** [comment]
Thanks, @dmehring. I went through updated notebook. It now contains a caveat on the use of `where`. That will help users (including CASA developers) to understand the difference between `where` and other selections like `isel`. And thank you for adding another cell to manipulate pacman image as a casa image. That worked to me. I tried to open pacman.im with CARTA, and it was successful without any tweak to the image/image header. I think we can close this ticket now. Thanks again @dmehring for the improvement of notebook!
